### PR TITLE
fix for Math.round(4503599627370497) and Math.round(0.49999999999999994)

### PR DESCRIPTION
**Pull request by Yaffle**

### CODE REVIEW

File: lib/Runtime/Library/MathLibrary.cpp [hunk 1]
```diff
@@ -1030,6 +1030,14 @@ namespace Js
             {
                 return scriptContext->GetLibrary()->GetNegativeZero();
             }
+            if(x <= -(1 / std::numeric_limits<double>::epsilon() + 1) || x >= (1 / std::numeric_limits<double>::epsilon() + 1))
```
Comment (MikeHolman) [inline lib/Runtime/Library/MathLibrary.cpp]: can you use Math::EPSILON? std::numeric_limits doesn't build.

Comment (Yaffle) [inline lib/Runtime/Library/MathLibrary.cpp]: ups... OK

**Comment by MikeHolman** [comment]
This will not fix issue in JIT either. Can you look at how code is generated for InlineMathRound in LowerMDShared.cpp (around line 8670) and GenerateFastInlineBuiltInMathRound in arm\LowerMD.cpp?

Also, would be great to have a UT for this.

**Comment by Yaffle** [comment]
will look at it later

**Comment by abchatra** [comment]
Any update here?

**Comment by Yaffle** [comment]
@abchatra , no, sorry, I have looked at code generation and understood that this is not so easy (for me)

**Comment by abchatra** [comment]
@Yaffle Sure no problem. Can you open an issue for this so someone else can pick up? Please provide a concise repro. I will close this PR for now. If you come back to it we can always reopen it.

**Comment by Yaffle** [comment]
@abchatra https://github.com/Microsoft/ChakraCore/issues/178

**Comment by mcfi** [comment]
@Yaffle 's change seems a bit hacky, and below is perhaps a more general way of fixing the code. It passes all tests on https://github.com/tc39/test262/blob/26e6fd7c1779a63913cc7720cbc6c87b3b7b3285/test/built-ins/Math/round/S15.8.2.15_A7.js

```c++
diff --git a/lib/Runtime/Library/MathLibrary.cpp b/lib/Runtime/Library/MathLibrary.cpp
index 0a285cc..898fd31 100644
--- a/lib/Runtime/Library/MathLibrary.cpp
+++ b/lib/Runtime/Library/MathLibrary.cpp
             {
                 return scriptContext->GetLibrary()->GetNaN();
             }
-            if(x == 0.0 || !NumberUtilities::IsFinite(x))
+
+            // x is positive or negative infinity, or an integer, or negative zero
+            if (!NumberUtilities::IsFinite(x) || x == ::floor(x))
             {
-                // 0.0 catches the -0 case...
                 return JavascriptNumber::ToVarNoCheck(x, scriptContext);
             }

+            if (x > 0 && x < 0.5)
+            {
+                return JavascriptNumber::ToVarNoCheck((double)Js::JavascriptNumber::k_Zero, scriptContext);
+            }
+
             if(x < 0 && x >= -0.5)
             {
                 return scriptContext->GetLibrary()->GetNegativeZero();
```

Notice, that the Math::Round uses the ```floor``` function, the JIT code needs to be revised accordingly. Anyone could explain how to make JITted code call a native function, or any equivalent way?

**Comment by MikeHolman** [comment]
You can call a native function using `m_lowererMD.ChangeToHelperCall`, and load arguments using `m_lowererMD.Load[Double]HelperArgument`. However, in JIT code we want to use machine instructions over calls whenever possible, e.g. in this case `ROUNDSD` instruction on x86/x64 (it can be used in place of floor function). Worth keeping the [Intel Reference Manual](http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf) handy when doing these types of changes. 

And note that this can be perf sensitive code (especially in JIT code), so we want to make sure our algorithm is performant. I don't know what the exact guidelines are for perf testing, but at the least you should verify that perf is flat across the benchmarks we have in test folder. We have a perl script for running them.

**Comment by mcfi** [comment]
@MikeHolman , ok, I'll try to rewrite the JIT code generator using machine instructions that basically lower the above changes to MathLibrary.cpp.

**Pull request closed without merging**